pyteal/ast/itxn.py

```diff
@@ -200,6 +205,108 @@ def SetFields(cls, fields: Dict[TxnField, Union[Expr, List[Expr]]]) -> Expr:
+    ) -> Expr:
+
+        # Default, always need these
+        fields_to_set = [
+            cls.SetField(TxnField.type_enum, TxnType.ApplicationCall),
+            cls.SetField(TxnField.application_id, app_id),
+        ]
+
+        # We only care about the args
+        arg_type_specs: list[abi.TypeSpec]
+        arg_type_specs, _ = type_specs_from_signature(method_signature)
+
+        # Start app args with the method selector
+        app_args: list[Expr] = [MethodSignature(method_signature)]
+
+        # Transactions are not included in the App Call
+        txns_to_pass: list[Expr] = []
+
+        # Reference Types are treated specially
+        accts: list[Expr] = []
+        apps: list[Expr] = []
+        assets: list[Expr] = []
+
+        for idx, arg_ts in enumerate(arg_type_specs):
+            arg = args[idx]
+
+            if arg_ts in abi.TransactionTypeSpecs:
+                if not isinstance(arg, dict):
+                    raise TealTypeError(arg, dict[TxnField, Expr | list[Expr]])
+
+                txns_to_pass.append(InnerTxnBuilder.SetFields(arg))
+
+            elif arg_ts in abi.ReferenceTypeSpecs:
+                match arg_ts:
+                    # For both acct and application, add index to
+                    # app args _after_ appending since 0 is implicitly set
+                    case abi.AccountTypeSpec():
+                        if isinstance(arg, Expr):
+                            accts.append(arg)
+                        elif isinstance(arg, abi.Account):
+                            accts.append(arg.address())
+                        else:
+                            raise TealTypeError(arg, abi.Account | Expr)
+
+                        app_args.append(uint_encode(8, Int(len(accts))))
+
+                    case abi.ApplicationTypeSpec():
+                        if isinstance(arg, Expr):
+                            apps.append(arg)
+                        elif isinstance(arg, abi.Application):
+                            apps.append(arg.application_id())
+                        else:
+                            raise TealTypeError(arg, abi.Application | Expr)
+
+                        app_args.append(uint_encode(8, Int(len(apps))))
+
+                    # For assets, add to app_args prior to appending to assets array
+                    case abi.AssetTypeSpec():
+                        app_args.append(uint_encode(8, Int(len(assets))))
+
+                        if isinstance(arg, Expr):
+                            assets.append(arg)
+                        elif isinstance(arg, abi.Asset):
+                            assets.append(arg.asset_id())
+                        else:
+                            raise TealTypeError(arg, abi.Asset | Expr)
+            else:
+                if isinstance(arg, Expr):
+                    app_args.append(arg)
+                elif isinstance(arg, abi.BaseType):
+                    app_args.append(arg.encode())
```

Comment:
I'd like to see more type checking here, specifically making sure `arg`'s type spec matches `arg_ts`